homeassistant/components/rachio/calendar.py

```diff
@@ -0,0 +1,180 @@
+"""Rachio smart hose timer calendar."""
+
+from datetime import datetime, timedelta
+import logging
+from typing import Any
+
+from homeassistant.components.calendar import (
+    CalendarEntity,
+    CalendarEntityFeature,
+    CalendarEvent,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    DOMAIN as DOMAIN_RACHIO,
+    KEY_ADDRESS,
+    KEY_DURATION_SECONDS,
+    KEY_ID,
+    KEY_LOCALITY,
+    KEY_PROGRAM_ID,
+    KEY_PROGRAM_NAME,
+    KEY_RUN_SUMMARIES,
+    KEY_SERIAL_NUMBER,
+    KEY_SKIP,
+    KEY_SKIPPABLE,
+    KEY_START_TIME,
+    KEY_TOTAL_RUN_DURATION,
+    KEY_VALVE_NAME,
+)
+from .coordinator import RachioScheduleUpdateCoordinator
+from .device import RachioPerson
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up entry for Rachio smart hose timer calendar."""
+    person: RachioPerson = hass.data[DOMAIN_RACHIO][config_entry.entry_id]
+    async_add_entities(
+        RachioCalendarEntity(base_station.schedule_coordinator, base_station)
+        for base_station in person.base_stations
+    )
+
+
+class RachioCalendarEntity(
+    CoordinatorEntity[RachioScheduleUpdateCoordinator], CalendarEntity
+):
+    """Rachio calendar entity."""
+
+    _attr_has_entity_name = True
+    _attr_translation_key = "calendar"
+    _attr_supported_features = CalendarEntityFeature.DELETE_EVENT
+
+    def __init__(
+        self, coordinator: RachioScheduleUpdateCoordinator, base_station
+    ) -> None:
+        """Initialize a Rachio calendar entity."""
+        super().__init__(coordinator)
+        self.base_station = base_station
+        self._event: CalendarEvent | None = None
+        self._location = coordinator.base_station[KEY_ADDRESS][KEY_LOCALITY]
+        self._attr_translation_placeholders = {
+            "base": coordinator.base_station[KEY_SERIAL_NUMBER]
+        }
+        self._attr_unique_id = f"{coordinator.base_station[KEY_ID]}-calendar"
+        self._previous_event: dict[str, Any] | None = None
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        event = self._handle_upcoming_event()
+        if not event:
```

Comment:
```suggestion
        if not (event := self._handle_upcoming_event()):
```

preference here, but I'd probably write it like this